File: src/batch_scheduler.cpp
#include "stator/batch_scheduler.hpp"
#include "stator/gmsh_backend.hpp"
#include "stator/geometry_builder.hpp"
#include "stator/topology_registry.hpp"
#include "stator/mesh_generator.hpp"
#include "stator/export_engine.hpp"
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <thread>
#include <chrono>
#include <filesystem>

// POSIX
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>
#include <unistd.h>

namespace stator {

// ─── execute_job ──────────────────────────────────────────────────────────────
// Runs inside the child process. Returns 0 on success.

int BatchScheduler::execute_job(const BatchJob& job, const std::string& status_path) {
    std::string error_msg;
    std::string msh_path, vtk_path, hdf5_path, json_path;
    bool success = false;

    try {
        // Step 1: validate params
        BatchJob mutable_job = job;
        mutable_job.params.validate_and_derive();

        // Step 2: backend
        auto backend = make_default_backend();

        // Step 3: initialize
        std::string stem = ExportEngine::compute_stem(mutable_job.params);
        backend->initialize("stator_" + stem);

        // Step 4: geometry
        GeometryBuilder builder(backend.get());
        auto geo = builder.build(mutable_job.params);
        if (!geo.success)
            throw std::runtime_error("Geometry build failed: " + geo.error_message);

        // Step 5-6: topology + winding
        TopologyRegistry registry(mutable_job.params.n_slots);
        // (physical group assignment happens inside MeshGenerator::generate)

        // Step 7-8: mesh
        MeshGenerator mesher(backend.get(), mutable_job.mesh_config);
        auto mesh_result = mesher.generate(mutable_job.params, geo, registry);
        if (!mesh_result.success)
            throw std::runtime_error("Mesh generation failed: " + mesh_result.error_message);

        // Steps 9-10: export
        namespace fs = std::filesystem;
        fs::create_directories(mutable_job.export_config.output_dir);
        ExportEngine exporter(backend.get());
        auto export_results = exporter.write_all_sync(
            mutable_job.params, mesh_result, mutable_job.export_config);

        std::string base = mutable_job.export_config.output_dir + "/" + stem;
        for (auto& er : export_results) {
            if (!er.success) continue;
            if (er.format == ExportFormat::MSH)  msh_path  = er.path;
            if (er.format == ExportFormat::VTK)  vtk_path  = er.path;
            if (er.format == ExportFormat::HDF5) hdf5_path = er.path;
            if (er.format == ExportFormat::JSON) json_path = er.path;
        }

        // Step 11: finalize
        backend->finalize();
        success = true;

    } catch (const std::exception& e) {
        error_msg = e.what();
    } catch (...) {
        error_msg = "Unknown exception in execute_job";
    }

    // Write status JSON
    try {
        std::ofstream f(status_path);
        f << "{"
          << "\"job_id\":\"" << job.job_id << "\","
          << "\"success\":" << (success ? "true" : "false") << ","
          << "\"error\":\"" << error_msg << "\","
          << "\"msh_path\":\"" << msh_path << "\","
          << "\"vtk_path\":\"" << vtk_path << "\","
          << "\"hdf5_path\":\"" << hdf5_path << "\","
          << "\"json_path\":\"" << json_path << "\""
          << "}";
    } catch (...) {
        // Cannot write status — fall through
    }

    return success ? 0 : 1;
}

// ─── read_status_json ─────────────────────────────────────────────────────────

BatchResult BatchScheduler::read_status_json(const std::string& path,
                                              const std::string& job_id) {
    BatchResult r;
    r.job_id = job_id;
    try {
        std::ifstream f(path);
        if (!f) { r.error = "Cannot open status file: " + path; return r; }
        std::string content((std::istreambuf_iterator<char>(f)),
                             std::istreambuf_iterator<char>());

        // Minimal JSON field extraction (no external JSON lib)
        auto extract = [&](const std::string& key) -> std::string {
            std::string search = "\"" + key + "\":\"";
            auto pos = content.find(search);
            if (pos == std::string::npos) return "";
            pos += search.size();
            auto end = content.find('"', pos);
            return content.substr(pos, end - pos);
        };
        auto extract_bool = [&](const std::string& key) -> bool {
            std::string search = "\"" + key + "\":";
            auto pos = content.find(search);
            if (pos == std::string::npos) return false;
            pos += search.size();
            return content.substr(pos, 4) == "true";
        };

        r.success   = extract_bool("success");
        r.error     = extract("error");
        r.msh_path  = extract("msh_path");
        r.vtk_path  = extract("vtk_path");
        r.hdf5_path = extract("hdf5_path");
        r.json_path = extract("json_path");
    } catch (const std::exception& e) {
        r.error = e.what();
    }
    return r;
}

// ─── write_summary ────────────────────────────────────────────────────────────

void BatchScheduler::write_summary(const std::string& output_dir,
                                    const std::vector<BatchResult>& results) {
    try {
        namespace fs = std::filesystem;
        fs::create_directories(output_dir);
        std::ofstream f(output_dir + "/batch_summary.json");
        f << "[";
        for (size_t i = 0; i < results.size(); ++i) {
            if (i) f << ",";
            const auto& r = results[i];
            f << "{"
              << "\"job_id\":\"" << r.job_id << "\","
              << "\"success\":" << (r.success ? "true" : "false") << ","
              << "\"error\":\"" << r.error << "\","
              << "\"msh_path\":\"" << r.msh_path << "\""
              << "}";
        }
        f << "]";
    } catch (...) {}
}

// ─── cancel ───────────────────────────────────────────────────────────────────

void BatchScheduler::cancel() {
    cancel_flag_.store(true);
}

// ─── run ──────────────────────────────────────────────────────────────────────

std::vector<BatchResult> BatchScheduler::run(const std::vector<BatchJob>& jobs,
                                              const BatchSchedulerConfig& config) {
    if (jobs.empty()) return {};

    running_.store(true);
    cancel_flag_.store(false);

    int max_par = config.max_parallel;
    if (max_par <= 0)
        max_par = static_cast<int>(std::thread::hardware_concurrency());
    if (max_par <= 0) max_par = 1;

    std::vector<BatchResult> results(jobs.size());
    std::string tmp_dir = std::filesystem::temp_directory_path().string() + "/stator_batch";
    std::filesystem::create_directories(tmp_dir);

    // Active child bookkeeping: (pid, job_index, start_time)
    struct ActiveChild {
        pid_t pid;
        int   job_idx;
        std::chrono::steady_clock::time_point start;
    };
    std::vector<ActiveChild> active;

    int jobs_done = 0;
    int n_jobs    = static_cast<int>(jobs.size());

    auto reap_one = [&](bool blocking) {
        for (auto it = active.begin(); it != active.end(); ) {
            // Check timeout
            if (config.job_timeout_sec > 0) {
                auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
                    std::chrono::steady_clock::now() - it->start).count();
                if (elapsed >= config.job_timeout_sec)
                    kill(it->pid, SIGKILL);
            }
            int status = 0;
            pid_t ret = waitpid(it->pid, &status, blocking ? 0 : WNOHANG);
            if (ret == it->pid) {
                // Child finished — read status JSON
                std::string sfile = tmp_dir + "/job_" + std::to_string(it->job_idx) + ".json";
                results[it->job_idx] = read_status_json(sfile, jobs[it->job_idx].job_id);
                results[it->job_idx].job_id = jobs[it->job_idx].job_id;
                ++jobs_done;
                if (progress_cb_)
                    progress_cb_(jobs_done, n_jobs,
                                  results[it->job_idx].success,
                                  results[it->job_idx].job_id);
                it = active.erase(it);
                return true;
            } else {
                ++it;
            }
        }
        return false;
    };

    for (int i = 0; i < n_jobs; ++i) {
        if (cancel_flag_.load()) {
            results[i].job_id = jobs[i].job_id;
            results[i].error  = "cancelled";
            continue;
        }

        // Check skip_existing
        if (config.skip_existing &&
            ExportEngine::outputs_exist(jobs[i].params, jobs[i].export_config)) {
            results[i].job_id = jobs[i].job_id;
            results[i].success = true;
            ++jobs_done;
            if (progress_cb_)
                progress_cb_(jobs_done, n_jobs, true, jobs[i].job_id);
            continue;
        }

        // Wait for a slot if at capacity
        while (static_cast<int>(active.size()) >= max_par) {
            reap_one(false);
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        std::string status_path = tmp_dir + "/job_" + std::to_string(i) + ".json";
        pid_t pid = fork();
        if (pid == 0) {
            // Child process
            int rc = execute_job(jobs[i], status_path);
            _exit(rc);
        } else if (pid > 0) {
            active.push_back({pid, i, std::chrono::steady_clock::now()});
        } else {
            results[i].job_id = jobs[i].job_id;
            results[i].error  = "fork() failed";
        }
    }

    // Wait for all remaining children
    while (!active.empty()) {
        if (!reap_one(false))
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Send SIGTERM to any children still active if cancelled
    for (auto& ac : active)
        kill(ac.pid, SIGTERM);

    // Write summary
    if (config.write_summary && !jobs.empty()) {
        std::string out_dir = jobs[0].export_config.output_dir;
        write_summary(out_dir, results);
    }

    running_.store(false);
    return results;
}

} // namespace stator

File: stator_pipeline/batch_scheduler.py
"""batch_scheduler.py — Pure Python batch scheduler using multiprocessing."""
from __future__ import annotations

import json
import os
from concurrent.futures import ProcessPoolExecutor, as_completed
from dataclasses import dataclass
from typing import Callable, Optional

from .params import StatorParams
from .export_engine import ExportConfig, ExportFormat, outputs_exist


@dataclass
class BatchJob:
    job_id: str
    params: StatorParams
    export_config: ExportConfig


@dataclass
class BatchConfig:
    max_parallel: int = 0
    skip_existing: bool = True
    job_timeout_sec: int = 300
    write_summary: bool = True


@dataclass
class BatchResult:
    job_id: str = ""
    success: bool = False
    error: str = ""
    msh_path: str = ""
    vtk_path: str = ""
    hdf5_path: str = ""
    json_path: str = ""


def _execute_job(job: BatchJob) -> BatchResult:
    """Run one job in a subprocess. Returns BatchResult."""
    from .gmsh_backend import make_default_backend
    from .geometry_builder import GeometryBuilder
    from .topology_registry import TopologyRegistry
    from .mesh_generator import MeshGenerator
    from .export_engine import ExportEngine, compute_stem

    result = BatchResult(job_id=job.job_id)
    try:
        backend = make_default_backend()
        stem = compute_stem(job.params)
        backend.initialize(f"stator_{stem}")

        builder = GeometryBuilder(backend)
        geo = builder.build(job.params)
        if not geo.success:
            result.error = f"Geometry build failed: {geo.error_message}"
            return result

        registry = TopologyRegistry(job.params.n_slots)
        generator = MeshGenerator(backend)
        mesh = generator.generate(job.params, geo, registry)
        if not mesh.success:
            result.error = f"Mesh generation failed: {mesh.error_message}"
            return result

        os.makedirs(job.export_config.output_dir, exist_ok=True)
        engine = ExportEngine(backend)
        export_results = engine.write_all(job.params, mesh, job.export_config)

        backend.finalize()

        for r in export_results:
            if not r.success:
                continue
            if ExportFormat.MSH in r.format:
                result.msh_path = r.path
            elif ExportFormat.VTK in r.format:
                result.vtk_path = r.path
            elif ExportFormat.HDF5 in r.format:
                result.hdf5_path = r.path
            elif ExportFormat.JSON in r.format:
                result.json_path = r.path

        result.success = True
    except Exception as e:
        result.error = str(e)
    return result


class BatchScheduler:
    def __init__(self) -> None:
        self._cancel = False

    def cancel(self) -> None:
        self._cancel = True

    def run(
        self,
        jobs: list[BatchJob],
        config: BatchConfig,
        progress_callback: Optional[Callable[[int, int, bool, str], None]] = None,
    ) -> list[BatchResult]:
        if not jobs:
            return []
        self._cancel = False

        max_workers = config.max_parallel or os.cpu_count() or 1
        results: list[BatchResult] = []
        jobs_done = 0
        total = len(jobs)

        # Split into skip-existing and to-run
        run_jobs: list[BatchJob] = []
        for job in jobs:
            if config.skip_existing and outputs_exist(job.params, job.export_config):
                r = BatchResult(job_id=job.job_id, success=True)
                results.append(r)
                jobs_done += 1
                if progress_callback:
                    progress_callback(jobs_done, total, True, job.job_id)
            else:
                run_jobs.append(job)

        with ProcessPoolExecutor(max_workers=max_workers) as executor:
            future_to_job = {executor.submit(_execute_job, j): j for j in run_jobs}
            for future in as_completed(future_to_job):
                if self._cancel:
                    break
                r = future.result()
                results.append(r)
                jobs_done += 1
                if progress_callback:
                    progress_callback(jobs_done, total, r.success, r.job_id)

        if config.write_summary and jobs:
            output_dir = jobs[0].export_config.output_dir
            os.makedirs(output_dir, exist_ok=True)
            summary_path = os.path.join(output_dir, "batch_summary.json")
            with open(summary_path, "w") as f:
                json.dump([
                    {"job_id": r.job_id, "success": r.success, "error": r.error, "msh_path": r.msh_path}
                    for r in results
                ], f, indent=2)

        return results
